crates/sui-core/src/authority_server.rs

```diff
@@ -485,72 +557,103 @@ impl ValidatorService {
             // 3) All certificates are sent to consensus (at least by some authorities)
             // For shared objects this will wait until either timeout or we have heard back from consensus.
             // For owned objects this will return without waiting for certificate to be sequenced
-            // First do quick dirty non-async check
-            if !epoch_store.is_tx_cert_consensus_message_processed(&certificate)? {
+            // First do quick dirty non-async check.
+            //
+            // For Soft Bundle, if at this point we know at least one certificate has already been processed,
+            // reject the entire bundle.  Otherwise, submit all certificates in one request.
+            let should_submit = if is_soft_bundle {
+                fp_ensure!(
+                    !epoch_store
+                        .is_any_tx_certs_consensus_message_processed(&verified_certificates)?,
```

Comment:
If you move this check to be together with all the other soft bundle specific checks, here you can simply submit if any cert is not already submitted. Once you do so, we could then even remove the is_soft_bundle argument from this function.

Reply:
I have moved the check to `soft_bundle_validity_check`, which is closer to its RPC handler.
Now the only place that relies on `is_soft_bundle` is where we determine whether it is necessary to check the first certificate has been executed.